### PR TITLE
KEYCLOAK-17685 - Add a javascript identity provider mapper

### PR DESCRIPTION
**Pull request by laskasn**
Currently, keycloak has a script mapper only on the Client side. This type of mapper is very useful when you want to perform a customized mapping, not covered from the other mappings. For instance, if you want to perform a conditional mapping - if this, map using attribA, otherwise map using attribB. or want to have a mapping with perplex concatenations, etc. We have implemented the same "script mapper" for the IdentityProvider side. 
We use this mapper in our installations, with javascript code to perform such complex mapping scenarios, which cannot be done using any of the existing mapper(s).
We think it's a very handy tool and we would like to contribute this to the community.

### CODE REVIEW

**Comment by laskasn** [comment]
Oops, i just saw that this PR was flagged "missing/tests" and was put on hold. 
I now added the required tests and rebased to the master. I think now the PR is ok, do you think we could resume it @hmlnarik ?

File: services/src/main/java/org/keycloak/broker/provider/ScriptMapper.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/broker/provider/ScriptMapper.java:5]: Minor: Will be good to remove star imports and import all classes separately (See other files of Keycloak codebase).

File: services/src/main/java/org/keycloak/broker/provider/ScriptMapper.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/broker/provider/ScriptMapper.java:10]: Same comment as before

File: services/src/main/java/org/keycloak/broker/provider/ScriptMapper.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/broker/provider/ScriptMapper.java:38]: userModel should be also added here?

**Review by mposolda — COMMENTED** [review]
@laskasn CC @cgeorgilakis Sorry for late response.

The PR looks good to me and functionality would be nice addition to Keycloak! Also Thanks for adding the tests.

I've added few comments inline. Besides that, there are some updates to script providers since you created this PR. For security reasons, we don't expect anymore javascript providers directly update-able by administrator in admin console. Instead of this, we support only javascript providers, which are deployed to the server in the `providers` directory of the distribution. See documentation for more details: https://www.keycloak.org/docs/latest/server_development/index.html#_script_providers . 

See for instance `DeployedScriptAuthenticatorFactory` or `DeployedScriptOIDCProtocolMapper` and various places where those are used. Also see this commit for the inspiration of the various places, which might need to be updated in this PR for support this as deployed provider https://github.com/keycloak/keycloak/commit/040e52cf (As the Deployed SAML protocol mapper support was added as a follow-up).

Finally, the docs would need to be added (This is not yet present in the commit above due the keycloak-documentation was in different repository, but now it is merged to the main repository). I think some updates may be needed to this page: https://www.keycloak.org/docs/latest/server_development/index.html#_script_providers .

File: services/src/main/java/org/keycloak/broker/provider/ScriptMapper.java
Comment (cgeorgilakis) [reply services/src/main/java/org/keycloak/broker/provider/ScriptMapper.java]: Done

File: services/src/main/java/org/keycloak/broker/provider/ScriptMapper.java
Comment (cgeorgilakis) [reply services/src/main/java/org/keycloak/broker/provider/ScriptMapper.java]: Yes. Added.

**Comment by cgeorgilakis** [comment]
> @laskasn CC @cgeorgilakis Sorry for late response.
> 
> The PR looks good to me and functionality would be nice addition to Keycloak! Also Thanks for adding the tests.
> 
> I've added few comments inline. Besides that, there are some updates to script providers since you created this PR. For security reasons, we don't expect anymore javascript providers directly update-able by administrator in admin console. Instead of this, we support only javascript providers, which are deployed to the server in the `providers` directory of the distribution. See documentation for more details: https://www.keycloak.org/docs/latest/server_development/index.html#_script_providers .
> 
> See for instance `DeployedScriptAuthenticatorFactory` or `DeployedScriptOIDCProtocolMapper` and various places where those are used. Also see this commit for the inspiration of the various places, which might need to be updated in this PR for support this as deployed provider [040e52cf](https://github.com/keycloak/keycloak/commit/040e52cf) (As the Deployed SAML protocol mapper support was added as a follow-up).
> 
> Finally, the docs would need to be added (This is not yet present in the commit above due the keycloak-documentation was in different repository, but now it is merged to the main repository). I think some updates may be needed to this page: https://www.keycloak.org/docs/latest/server_development/index.html#_script_providers .

@mposolda thanks for the information. I have started implementing needed changes (see my last commit) for making this mapper only available through [script providers](https://www.keycloak.org/docs/latest/server_development/index.html#_script_providers). However, I do not understand how you can map the deployed IdP script mapper to a SAML Identity Provider. Could you help?

**Comment by jonkoops** [comment]
@mposolda looks like this PR needs another review, could you take a look at this?

**Review by mposolda — APPROVED** [review]
@laskasn @cgeorgilakis Thanks and sorry for late response! I hope to merge if tests are OK.

Is it also possible to add documentation in this PR? I belive that this and failing tests might be the only parts needed before this PR can be merged. think this section will need some update: https://www.keycloak.org/docs/23.0.6/server_development/#_script_providers . Also can be nice to add some very small note here https://www.keycloak.org/docs/latest/server_admin/index.html#_mappers with the link to the https://www.keycloak.org/docs/23.0.6/server_development/#_script_providers documentation for more details (See this for the inspiration https://www.keycloak.org/docs/latest/server_admin/index.html#script-mapper )

**Comment by cgeorgilakis** [comment]
> > @laskasn CC @cgeorgilakis Sorry for late response.
> > The PR looks good to me and functionality would be nice addition to Keycloak! Also Thanks for adding the tests.
> > I've added few comments inline. Besides that, there are some updates to script providers since you created this PR. For security reasons, we don't expect anymore javascript providers directly update-able by administrator in admin console. Instead of this, we support only javascript providers, which are deployed to the server in the `providers` directory of the distribution. See documentation for more details: https://www.keycloak.org/docs/latest/server_development/index.html#_script_providers .
> > See for instance `DeployedScriptAuthenticatorFactory` or `DeployedScriptOIDCProtocolMapper` and various places where those are used. Also see this commit for the inspiration of the various places, which might need to be updated in this PR for support this as deployed provider [040e52cf](https://github.com/keycloak/keycloak/commit/040e52cf) (As the Deployed SAML protocol mapper support was added as a follow-up).
> > Finally, the docs would need to be added (This is not yet present in the commit above due the keycloak-documentation was in different repository, but now it is merged to the main repository). I think some updates may be needed to this page: https://www.keycloak.org/docs/latest/server_development/index.html#_script_providers .
> 
> @mposolda thanks for the information. I have started implementing needed changes (see my last commit) for making this mapper only available through [script providers](https://www.keycloak.org/docs/latest/server_development/index.html#_script_providers). However, I do not understand how you can map the deployed IdP script mapper to a SAML Identity Provider. Could you help?

@mposolda see my last comment and my last commit. I believe that this PR is now unfinished (related to needed changes mentioned by you).
Maybe test failures is related with this. Test logs are not available due to expiration.

**Comment by cgeorgilakis** [comment]
> @laskasn @cgeorgilakis Thanks and sorry for late response! I hope to merge if tests are OK.
> 
> Is it also possible to add documentation in this PR? I belive that this and failing tests might be the only parts needed before this PR can be merged. think this section will need some update: https://www.keycloak.org/docs/23.0.6/server_development/#_script_providers . Also can be nice to add some very small note here https://www.keycloak.org/docs/latest/server_admin/index.html#_mappers with the link to the https://www.keycloak.org/docs/23.0.6/server_development/#_script_providers documentation for more details (See this for the inspiration https://www.keycloak.org/docs/latest/server_admin/index.html#script-mapper )

I will see docs before merging.
However, I believe that we firstly need to finish PR code. See my previous comment.

**Comment by mposolda** [comment]
@cgeorgilakis Yes, true. From the top of my head, I am not sure what is missing and the exact reason of the test failure.

Hopefully you can investigate and figure it (based on existing tests etc). Maybe isolating the problem a bit can help. Are you able to manually deploy the provider to the Keycloak distribution based on your PR? If yes, it is clear that issue is somewhere in the tests. Otherwise it is in the server-code itself.

**Comment by mposolda** [comment]
@cgeorgilakis I am closing as the PR is open for a long time without any progress. Feel free to re-open when you figure the issues we discussed above.

**Pull request closed without merging**
